segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -164,12 +164,15 @@ public TableBasedMetadataStore(String tableName, TableStore tableStore, Executor
                             }, getExecutor())
                             .thenComposeAsync(v2 -> {
                                 // Delete deleted keys.
-                                return this.tableStore.remove(tableName, keysToDelete, timeout);
+                                return this.tableStore.remove(tableName, keysToDelete, timeout)
+                                        .handleAsync((v1, ex) -> {
+                                            if (ex == null) {
+                                                deletedKeyToTxnDataMap.values().stream().forEach(txnData -> txnData.setDbObject(TableKey.NOT_EXISTS));
+                                            }
+                                            return null;
+                                        }, getExecutor());
                             }, getExecutor())
                             .thenRunAsync(() -> {
-                                for (val deletedKey : keysToDelete) {
```

Comment:
I don't understand how does moving this for loop from here to the `handleAsync` resolve this problem. `thenRunAsync` is executed only on success, which is exactly what you're doing inside `handleAsync`. 

What am I missing?

Reply:
We delete keys in two steps
1. Update them by marking them deleted in the value part of K-V pair. Note that we update all updated keys as well as keys marked deleted in a single transaction. After this step version of all keys changes. Record new version. 
2. Delete the "to be deleted" keys by calling remove().

This change deals when step 1  succeeds and 2 fails.
It used to fail the whole commit, but the version of all keys were changed.
What this commit does is that it basically ignores the failure in step 2.  TableBasedMetadata correctly handles when key is found marked deleted.

Reply:
- This has to be done this way only because that table segment cannot delete and update in the same transaction.
- Now, each SLTS metadata is a key-value pair that has `key -> { version, dbVersion, value, flags}`  structure. The version here is not same as dbVersion because of lazy committing.  However every time we write the table segment the dbVersion is used and updated.
- Before each commit the in memory and cached data is validated to make sure transaction does not use stale version. This is done before this writeAll method is called.
- The way delete is done is by first committing an update that sets the value to null. Then as a second step remove the KV pair. First step is conditional update, while the second step is unconditional delete.
- In case when update succeeds but remove fails , the value is already set to null which is interpreted as "not exists". As far as logic is concerned the key is "marked deleted" and that info is safely persisted already.
- Failure to remove should not fail the transaction. The dbVersion of keys correctly match the version returned by table segment. So if in future this key is again modified and committed we'll not have the version mismatch exception (The original bug that this fix is trying to solve).

Reply:
@sachin-j-joshi Let me rewrite your changes using abstract symbols.

Case 1 (Before):
```
A
.thenApply(B)
.thenCompose(C)
.thenRun(D)
```

Case 2 (After):
```
A
.thenCompose(B & C.handle( 
                       (no-error)->D; 
		                error-> ...))
```
		 
Situations:

A fails:
- 1: B, C, D not executed
- 2: B, C, D not executed

A succeeds, B Fails:
- 1: C, D not executed
- 2: C, D not executed

A, B succeed, C fails:
- 1: D not executed
- 2: D not executed (since you have `ex == null`)

A, B, C succeed:
1: D executes
2: D executes

That's why @RaulGracia  and I say we don't understand how this change could possibly have altered the program behavior. You reduced the number of CompletableFutures by combining B and C together, but that doesn't alter the behavior.